### PR TITLE
 Add a special class if a TreeView parent node has a selected child

### PR DESCRIPTION
**Pull request by alexander-kotov-dx**

### CODE REVIEW

**Review by DokaRus — DISMISSED** [review]

File: js/ui/hierarchical_collection/ui.data_adapter.js [hunk 1]
```diff
@@ -51,6 +52,7 @@ var DataAdapter = Class.inherit({
             searchMode: "contains",
             dataConverter: new HierarchicalDataConverter(),
             onNodeChanged: commonUtils.noop,
+            hasSelectedChild: false,
```
Comment (mrjohn0011) [inline js/ui/hierarchical_collection/ui.data_adapter.js]: Why we should use a feature toggle option? There is no breaking change in this feature. This option should be removed

File: js/ui/hierarchical_collection/ui.data_adapter.js [hunk 2]
```diff
@@ -139,7 +141,13 @@ var DataAdapter = Class.inherit({
             if(parent && node.internalFields.parentKey !== that.options.rootValue) {
                 that._iterateParents(node, function(parent) {
                     var newParentState = that._calculateSelectedState(parent);
-                    that._setFieldState(parent, SELECTED, newParentState);
+                    if(!withoutSelection) {
+                        that._setFieldState(parent, SELECTED, newParentState);
+                    }
+
+                    if(node.internalFields.hasOwnProperty(HASSELECTEDCHILD)) {
```
Comment (mrjohn0011) [inline js/ui/hierarchical_collection/ui.data_adapter.js]: Don't use node property to assign a CSS class. Has-selection classes should be reassigned dynamically when the selection was changed

File: js/ui/hierarchical_collection/ui.data_adapter.js [hunk 3]
```diff
@@ -130,7 +132,7 @@ var DataAdapter = Class.inherit({
         });
     },
 
-    _setParentSelection: function() {
+    _setParentSelectionProperties: function(withoutSelection) {
```
Comment (mrjohn0011) [inline js/ui/hierarchical_collection/ui.data_adapter.js]: Don't use a magic boolean parameter. It is unclear what this function do when we call `_setParentSelectionProperties(true)`

File: js/ui/tree_view/ui.tree_view.base.js [hunk 4]
```diff
@@ -853,7 +854,7 @@ var TreeViewBase = HierarchicalCollectionWidget.inherit({
         showCheckBox && this._renderCheckBox($node, node);
 
         this.setAria("selected", nodeData.selected, $node);
-        this._toggleSelectedClass($node, nodeData.selected);
+        this._toggleSelectionClasses($node, nodeData.selected, nodeData.hasSelectedChild);
```
Comment (mrjohn0011) [inline js/ui/tree_view/ui.tree_view.base.js]: This is unclear to use 3 parameters in this method. If the method called `toggleSelectionClasses`, it should assign more than 2 classes. Change the method's signature

File: testing/tests/DevExpress.ui.widgets/treeViewParts/selection.js [hunk 5]
```diff
@@ -372,3 +372,29 @@ QUnit.test("items should be selectable after the search", function(assert) {
 
     assert.equal(clickHandler.callCount, 1, "click works");
 });
+
+var HAS_SELECTED_CHILD_CLASS = "dx-treeview-node-has-selected-items";
+
+QUnit.test("parent has a special class if node has selected child", function(assert) {
```
Comment (mrjohn0011) [inline testing/tests/DevExpress.ui.widgets/treeViewParts/selection.js:381]: More tests required:

1. Recursive selection is enabled, select a middle-level child. Check if sub-children and parents have this class
2. Select all
3. Recursive selection is enabled/Disabled. Select/Deselect branch
4. Virtual mode/Create children. Select a leaf, load children, check the state
5. Select a parent of the Invisible/disabled children
6. Selection state is assigned on init. Select a middle-level child with the recursive selection. Check the state. Add this test to the server-side rendering (markup) tests

**Review by mrjohn0011 — CHANGES_REQUESTED** [review]

File: js/ui/hierarchical_collection/ui.data_adapter.js [hunk 6]
```diff
@@ -13,7 +13,8 @@ var Class = require("../../core/class"),
 
 var EXPANDED = "expanded",
     SELECTED = "selected",
-    DISABLED = "disabled";
+    DISABLED = "disabled",
+    HASSELECTEDCHILD = "hasSelectedChild";
```
Comment (ovchinnikov) [inline js/ui/hierarchical_collection/ui.data_adapter.js]: I suggest HAS_SELECTED_CHILD name for variable name

**Review by ovchinnikov — CHANGES_REQUESTED** [review]

**Pull request closed without merging**